xdsl/dialects/experimental/math.py

```diff
@@ -0,0 +1,170 @@
+from __future__ import annotations
+
+from typing import Annotated, Union
+
+from xdsl.dialects.builtin import IntegerType, AnyFloat, Attribute
+from xdsl.ir import Operation, SSAValue, OpResult
+from xdsl.irdl import irdl_op_definition, OptOpAttr, Operand
+from xdsl.dialects.arith import FastMathFlagsAttr
+
+
+@irdl_op_definition
+class FPowIOp(Operation):
+    """
+    Syntax:
+    operation ::= ssa-id `=` `math.fpowi` ssa-use `,` ssa-use `:` type
+
+    The fpowi operation takes a `base` operand of floating point type
+    (i.e. scalar, tensor or vector) and a `power` operand of integer type
+    (also scalar, tensor or vector) and returns one result of the same type
+    as `base`. The result is `base` raised to the power of `power`.
+    The operation is elementwise for non-scalars, e.g.:
+
+    %v = math.fpowi %base, %power : vector<2xf32>, vector<2xi32
+
+    The result is a vector of:
+
+    [<math.fpowi %base[0], %power[0]>, <math.fpowi %base[1], %power[1]>]
+
+    Example:
+
+    // Scalar exponentiation.
+    %a = math.fpowi %base, %power : f64, i32
+    """
+    name: str = "math.fpowi"
+    fastmath: OptOpAttr[FastMathFlagsAttr]
+    lhs: Annotated[Operand, AnyFloat]
+    rhs: Annotated[Operand, IntegerType]
+    result: Annotated[OpResult, AnyFloat]
+
+    @staticmethod
+    def get(lhs: Union[Operation, SSAValue],
+            rhs: Union[Operation, SSAValue],
+            fastmath: FastMathFlagsAttr | None = None) -> FPowIOp:
+        attributes: dict[str, Attribute] = {}
+        if fastmath is not None:
+            attributes["fastmath"] = fastmath
+
+        lhs = SSAValue.get(lhs)
```

Comment:
Why not for rhs as well?

Reply:
I just did this for the typing to keep PyRight happy as we grab out the .type, I don't think we really need it apart from keeping PyRight happy? But happy to do for RHS too if that keeps it all the same?

Reply:
Ah I see, yeah they differ there. I think, this code is bound to break at some point in the future, when we make build more restrictive, but we can also handle the issue then, I don't mind.

Reply:
I've updated it to do this for RHS too so that they are the same